### PR TITLE
Implement missing ME2 cards, reworked can't block alone, new damage redirect protection

### PR DESCRIPTION
**Pull request by ssk97**
~~Currently completely untested, and since some of these do some weird things they definitely need to be.~~ Also reworked CantBlockAloneAbility to not be special-cased in combat code.

edit: Added a few tests, I think this is good to merge but will leave it open for review for a bit

### CODE REVIEW

File: Mage.Sets/src/mage/cards/l/LavaBurst.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        if (targetPlayer != null) {
+            targetPlayer.damage(damage, source.getSourceId(), source, game, false, true);
+            return true;
+        }
+        Permanent targetPermanent = game.getPermanent(source.getFirstTarget());
+        if (targetPermanent != null) {
+            if (targetPermanent.isCreature(game)) {
+                // hack to avoid needing to create a DAMAGE_REDIRECTION event
+                // Pretend that all possible redirection effects have already been applied
+                List<UUID> redirectionEffects = game.getState().getContinuousEffects().getReplacementEffects().stream()
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/l/LavaBurst.java:76]: That's wrong implementation. Must be implemented same way as `damageEvent.isPreventable()` like `isRedirectable`. I recommend to remove that card from current PR.

File: Mage.Sets/src/mage/cards/s/SacredBoon.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+    @Override
+    public SacredBoonPreventDamageTargetEffect copy() {
+        return new SacredBoonPreventDamageTargetEffect(this);
+    }
+
+    @Override
+    public boolean replaceEvent(GameEvent event, Ability source, Game game) {
+        GameEvent preventEvent = new PreventDamageEvent(event.getTargetId(), source.getSourceId(), source, source.getControllerId(), event.getAmount(), ((DamageEvent) event).isCombatDamage());
+        if (!game.replaceEvent(preventEvent)) {
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/s/SacredBoon.java]: Why you don't use standard code with `PreventionEffectData preventionData = `? It's used in all other effects with same logic like Refraction Trap and many other cards.

File: Mage.Sets/src/mage/cards/s/SacredBoon.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            if (event.getAmount() >= this.amount) {
+                int damage = amount;
+                event.setAmount(event.getAmount() - amount);
+                this.used = true;
+                game.fireEvent(new PreventedDamageEvent(event.getTargetId(), source.getSourceId(), source, source.getControllerId(), damage));
+                prevented = damage;
+            } else {
+                int damage = event.getAmount();
+                event.setAmount(0);
+                amount -= damage;
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/s/SacredBoon.java]: for info - any damage related math must use overflow protection like overflowInc or overflowDec from CardUtils.

File: Mage.Sets/src/mage/cards/s/ScarsOfTheVeteran.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+    @Override
+    public ScarsOfTheVeteranPreventDamageTargetEffect copy() {
+        return new ScarsOfTheVeteranPreventDamageTargetEffect(this);
+    }
+
+    @Override
+    public boolean replaceEvent(GameEvent event, Ability source, Game game) {
+        GameEvent preventEvent = new PreventDamageEvent(event.getTargetId(), source.getSourceId(), source, source.getControllerId(), event.getAmount(), ((DamageEvent) event).isCombatDamage());
+        if (!game.replaceEvent(preventEvent)) {
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/s/ScarsOfTheVeteran.java]: see above about PreventionEffectData

File: Mage/src/main/java/mage/game/combat/CombatGroup.java [hunk 5]
```diff
@@ -679,29 +679,6 @@ public boolean checkBlockRestrictions(Game game, Player defender, int blockersCo
...
-        // effects: can't block alone
-        // too much blockers
-        if (blockersCount == 1) {
-            List<UUID> toBeRemoved = new ArrayList<>();
-            for (UUID blockerId : getBlockers()) {
-                Permanent blocker = game.getPermanent(blockerId);
-                if (blocker != null && blocker.getAbilities().containsKey(CantBlockAloneAbility.getInstance().getId())) {
-                    blockWasLegal = false;
-                    if (!game.isSimulation()) {
-                        game.informPlayers(blocker.getLogName() + " can't block alone. Removing it from combat.");
```
Comment (JayDi85) [inline Mage/src/main/java/mage/game/combat/CombatGroup.java:691]: Old logic informs about the problem in game logs (wrong block config). What about a new logic - is player can find the reason why he can't continue/block/attack by a creature?

Also deleted code was able to remove wrong blockers config -- what about a new code? Game must not able to save wrong block configuration.

**Review by JayDi85 — CHANGES_REQUESTED** [review]

File: Mage.Sets/src/mage/cards/s/SacredBoon.java (hunk 2 above)
Comment (ssk97) [reply Mage.Sets/src/mage/cards/s/SacredBoon.java]: These two were based on [[Test of Faith]], but you are right I think that all three should use that.

File: Mage.Sets/src/mage/cards/l/LavaBurst.java (hunk 1 above)
Comment (ssk97) [reply Mage.Sets/src/mage/cards/l/LavaBurst.java:76]: As shown in the test it does work, and while it definitely is a hack I can't think of any situation where it would fail. Why not include it? Whippoorwill (the only other card that can prevent redirection) would require a proper implementation, but this one does work without needing to mess with the fundamental damage code.

Comment (JayDi85) [reply Mage.Sets/src/mage/cards/l/LavaBurst.java:76]: Cause that code looks like: let’s remove other game effects, apply own changes and return it back.

File: Mage/src/main/java/mage/game/combat/CombatGroup.java (hunk 5 above)
Comment (ssk97) [reply Mage/src/main/java/mage/game/combat/CombatGroup.java:691]: It doesn't give a specific "can't block alone" message, but it does say that the specific creature "can't block this way".

I'm not sure exactly what you mean in the second comment. The game does not allow illegal blocks to happen, and there is the `CantAttackOrBlockAloneTest`.

File: Mage.Sets/src/mage/cards/l/LavaBurst.java (hunk 1 above)
Comment (ssk97) [reply Mage.Sets/src/mage/cards/l/LavaBurst.java:76]: Sure? What's wrong with that? It's not removing then returning them back, it doesn't modify the list at all. It just gets the replacement effects' IDs so that it can list them all in the "replacement effects that have already been applied" argument.

File: Mage/src/main/java/mage/game/combat/CombatGroup.java (hunk 5 above)
Comment (JayDi85) [reply Mage/src/main/java/mage/game/combat/CombatGroup.java:691]: All that restriction messages works in GUI only in most use cases and can’t be tested by unit tests. So it’s important to make sure in compatible logic for human player in real game. CantAttackOrBlockAloneTest and other uses non strict mode and testing AI logic - it cant catch wrong attack/block configurations like human player can try to do in real game.

File: Mage.Sets/src/mage/cards/s/SacredBoon.java (hunk 2 above)
Comment (ssk97) [reply Mage.Sets/src/mage/cards/s/SacredBoon.java]: Did some quick checking, will also correct Phyrexian Hydra, Brace for Impact, Temper, and Vigor.

**Comment by ssk97** [comment]
Welp, investigation shows there's some major issues in dozens of PreventionEffects. I think I'll probably split everything related to that into a separate PR, it's getting pretty huge.

_edit: Sorry for accidentally pushing to a new branch on mainline instead of here_

File: Mage.Sets/src/mage/cards/b/BondedHorncrest.java [hunk 6]
```diff
@@ -25,7 +26,7 @@ public BondedHorncrest(UUID ownerId, CardSetInfo setInfo) {
 
         // Bonded Horncrest can't attack or block alone.
         this.addAbility(new CantAttackAloneAbility());
-        this.addAbility(CantBlockAloneAbility.getInstance());
+        this.addAbility(new CantBlockAloneAbility());
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/b/BondedHorncrest.java]: With this ability no longer special cased, it should be possible to update "can't attack or block alone" to a single ability with two effects now to get the text right?

File: Mage.Sets/src/mage/cards/l/LavaBurst.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                        .filter(x -> x instanceof RedirectionEffect).map(Effect::getId).collect(Collectors.toList());
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/l/LavaBurst.java:77]: The main problem I see with this is some redirection effects aren't implemented as RedirectionEffect. See #10650. For this specific card, I have no issue with the approach of pretending all redirection effects have already been applied.

File: Mage.Sets/src/mage/cards/o/OrcishConscripts.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    public boolean canAttackCheckAfter(int numberOfAttackers, Ability source, Game game, boolean canUseChooseDialogs) {
+        return numberOfAttackers > 2;
+    }
+
+    @Override
+    public boolean applies(Permanent permanent, Ability source, Game game) {
+        return source.getSourceId().equals(permanent.getId());
+    }
+}
+class OrcishConscriptsBlockEffect extends RestrictionEffect {
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/o/OrcishConscripts.java:73]: I'd like to see tests to confirm that three of these can attack together and that three of these can block together.

File: Mage.Sets/src/mage/cards/o/OrcishFarmer.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        Permanent p = game.getPermanent(getTargetPointer().getFirst(game, source));
+        if (p == null) {
+            discard();
+            return false;
+        }
+        if (effectStartingOnTurn != game.getTurnNum()
+                && game.isActivePlayer(p.getControllerId())
+                && game.getStep().getType() == PhaseStep.UNTAP) {
+            discard();
+            return false;
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/o/OrcishFarmer.java:74]: This is weird, I would have expected to override `isInactive()`. But I guess this approach works.

File: Mage.Sets/src/mage/cards/s/SacredBoon.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    @Override
+    public boolean replaceEvent(GameEvent event, Ability source, Game game) {
+        PreventionEffectData preventionEffectData = preventDamageAction(event, source, game);
+        if (preventionEffectData.getPreventedDamage() > 0) {
+            Permanent targetPermanent = game.getPermanent(source.getTargets().getFirstTarget());
+            if (targetPermanent != null) {
+                DelayedTriggeredAbility delayedAbility = new AtTheBeginOfNextEndStepDelayedTriggeredAbility(
+                        new AddCountersTargetEffect(CounterType.P0P1.createInstance(preventionEffectData.getPreventedDamage()))
+                                .setTargetPointer(new FixedTarget(targetPermanent, game)));
+                game.addDelayedTriggeredAbility(delayedAbility, source);
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/s/SacredBoon.java:72]: don't suppose you have any ideas for making this technically correct for a single trigger from multiple replacement events? not required to fix but do add a comment noting it if it's not fixable

File: Mage.Tests/src/test/java/org/mage/test/cards/replacement/redirect/HarmsWayRedirectDamageTest.java [hunk 11]
```diff
@@ -132,4 +132,32 @@ public void testRedirectNotPreventableDamage() {
...
+        setChoice(playerA, "X=4");
+        castSpell(1, PhaseStep.PRECOMBAT_MAIN, playerB, "Harm's Way", playerA);
+        setChoice(playerB, "Lava Burst");
+
+        castSpell(1, PhaseStep.POSTCOMBAT_MAIN, playerA, "Lava Burst", playerB);
+        setChoice(playerA, "X=3");
+        castSpell(1, PhaseStep.POSTCOMBAT_MAIN, playerB, "Harm's Way", playerA);
+        setChoice(playerB, "Lava Burst");
+
+        setStopAt(1, PhaseStep.END_TURN);
```
Comment (xenohedron) [inline Mage.Tests/src/test/java/org/mage/test/cards/replacement/redirect/HarmsWayRedirectDamageTest.java:160]: strict choose mode

File: Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java [hunk 12]
```diff
@@ -55,13 +55,13 @@ A Characteristic Defining Ability (CDA) is an ability that defines a characteris
     protected boolean characterDefining = false;
 
     // until your next turn or until end of your next turn
-    private UUID startingControllerId; // player to check for turn duration (can't different with real controller ability)
-    private boolean startingTurnWasActive; // effect started during related players turn and related players turn was already active
-    private int effectStartingOnTurn = 0; // turn the effect started
-    private int effectStartingEndStep = 0;
-    private int nextTurnNumber = Integer.MAX_VALUE; // effect is waiting for a step during your next turn, we store it if found.
+    protected UUID startingControllerId; // player to check for turn duration (can't different with real controller ability)
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java]: I don't like exposing these, they aren't supposed to be modified. If you need some of them, why not add getter methods instead?

File: Mage/src/main/java/mage/abilities/keyword/CantBlockAloneAbility.java [hunk 13]
```diff
@@ -2,39 +2,26 @@
...
-
-    private Object readResolve() throws ObjectStreamException {
-        return instance;
-    }
+public class CantBlockAloneAbility extends SimpleStaticAbility {
 
-    public static CantBlockAloneAbility getInstance() {
-        return instance;
+    public CantBlockAloneAbility() {
+        super(Zone.BATTLEFIELD, new CantBlockAloneSourceEffect());
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/keyword/CantBlockAloneAbility.java]: I see CantAttackOrBlockAloneTest provides some coverage, though it's missing the case for two creatures with this ability

**Review by xenohedron — COMMENTED** [review]
For future, I would greatly prefer seeing independent changes in separate PRs.

File: Mage.Sets/src/mage/cards/s/SacredBoon.java (hunk 10 above)
Comment (ssk97) [reply Mage.Sets/src/mage/cards/s/SacredBoon.java:72]: Could maybe do something like storing the damage dealt amount inside `GameState.values` with a key derived from the MOR of the resolving spell, but I *really* don't like just throwing stuff into the global state like that.

File: Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java (hunk 12 above)
Comment (ssk97) [reply Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java]: Only really needed the `effectStartingOnTurn` for the `OrcishFarmer` discard timing, do you have a suggestion on a better way of doing "until the controller's next upkeep"? If not I'll just add the one getter method.

Comment (JayDi85) [reply Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java]: There are already standard methods like `isYourNextTurn` or `isYourNextUpkeepStep`

Comment (xenohedron) [reply Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java]: If an existing method works, by all means use it, but I'm pretty sure none of the existing methods are adequate here. Adding one getter method is fine

File: Mage.Sets/src/mage/cards/o/OrcishFarmer.java (hunk 9 above)
Comment (JayDi85) [reply Mage.Sets/src/mage/cards/o/OrcishFarmer.java:74]: effect's `discard()` usage for `Duration.Custom` is standard approach, just make sure it will be guaranty called in any use cases

File: Mage.Sets/src/mage/cards/s/SacredBoon.java (hunk 10 above)
Comment (JayDi85) [reply Mage.Sets/src/mage/cards/s/SacredBoon.java:72]: you can use `AddCountersTargetEffect` with custom `DynamicValue`, so total prevented value can be saved/loaded inside card's code only with `game.getState().getValue/setValue`

Comment (JayDi85) [reply Mage.Sets/src/mage/cards/s/SacredBoon.java:72]: Also you can ignore delay trigger creation on existing saved damage, so only 1 trigger will be created on multiple preventions.

File: Mage.Sets/src/mage/cards/o/OrcishConscripts.java (hunk 8 above)
Comment (ssk97) [reply Mage.Sets/src/mage/cards/o/OrcishConscripts.java:73]: A "Three of these blocking together" test is complicated by the fact that the test code tries to multi-block with the same Orcish Conscript and there's no way to specify which one.

**Review by xenohedron — APPROVED** [review]
Any improvement to single delayed trigger edge case can be a followup if desired

**Comment by JayDi85** [comment]
@xenohedron do not merge, effects removing hack is very bad

File: Mage.Sets/src/mage/cards/s/SacredBoon.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public boolean applies(GameEvent event, Ability source, Game game) {
+        if (!this.used && super.applies(event, source, game)) {
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/s/SacredBoon.java]: ah wait, do you actually want the `this.used` here?

**Comment by xenohedron** [comment]
Lava Burst can be discussed separately. Other cards merged via bf0f3bf, 227830d, bb845b3
